// amazon-textract-response-parser/dist/es/api-models/base.js
export var ApiRelationshipType;
(function (ApiRelationshipType) {
    ApiRelationshipType["Answer"] = "ANSWER";
    ApiRelationshipType["Child"] = "CHILD";
    ApiRelationshipType["ComplexFeatures"] = "COMPLEX_FEATURES";
    ApiRelationshipType["MergedCell"] = "MERGED_CELL";
    ApiRelationshipType["TableFooter"] = "TABLE_FOOTER";
    ApiRelationshipType["TableTitle"] = "TABLE_TITLE";
    ApiRelationshipType["Value"] = "VALUE";
})(ApiRelationshipType || (ApiRelationshipType = {}));
export var ApiBlockType;
(function (ApiBlockType) {
    ApiBlockType["Cell"] = "CELL";
    ApiBlockType["Key"] = "KEY";
    ApiBlockType["KeyValueSet"] = "KEY_VALUE_SET";
    ApiBlockType["LayoutFigure"] = "LAYOUT_FIGURE";
    ApiBlockType["LayoutFooter"] = "LAYOUT_FOOTER";
    ApiBlockType["LayoutHeader"] = "LAYOUT_HEADER";
    ApiBlockType["LayoutKeyValue"] = "LAYOUT_KEY_VALUE";
    ApiBlockType["LayoutList"] = "LAYOUT_LIST";
    ApiBlockType["LayoutPageNumber"] = "LAYOUT_PAGE_NUMBER";
    ApiBlockType["LayoutSectionHeader"] = "LAYOUT_SECTION_HEADER";
    ApiBlockType["LayoutTable"] = "LAYOUT_TABLE";
    ApiBlockType["LayoutText"] = "LAYOUT_TEXT";
    ApiBlockType["LayoutTitle"] = "LAYOUT_TITLE";
    ApiBlockType["Line"] = "LINE";
    ApiBlockType["MergedCell"] = "MERGED_CELL";
    ApiBlockType["Page"] = "PAGE";
    ApiBlockType["Query"] = "QUERY";
    ApiBlockType["QueryResult"] = "QUERY_RESULT";
    ApiBlockType["SelectionElement"] = "SELECTION_ELEMENT";
    ApiBlockType["Signature"] = "SIGNATURE";
    ApiBlockType["Table"] = "TABLE";
    ApiBlockType["TableFooter"] = "TABLE_FOOTER";
    ApiBlockType["TableTitle"] = "TABLE_TITLE";
    ApiBlockType["Value"] = "VALUE";
    ApiBlockType["Word"] = "WORD";
})(ApiBlockType || (ApiBlockType = {}));
const LAYOUT_BLOCK_TYPES = new Set([
    "LAYOUT_FIGURE",
    "LAYOUT_FOOTER",
    "LAYOUT_HEADER",
    "LAYOUT_KEY_VALUE",
    "LAYOUT_LIST",
    "LAYOUT_PAGE_NUMBER",
    "LAYOUT_SECTION_HEADER",
    "LAYOUT_TABLE",
    "LAYOUT_TEXT",
    "LAYOUT_TITLE",
]);
export function isLayoutBlockType(blockType) {
    return LAYOUT_BLOCK_TYPES.has(blockType);
}
