// amazon-textract-response-parser/dist/cjs/api-models/document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiTableCellEntityType = exports.ApiTableEntityType = exports.ApiKeyValueEntityType = exports.ApiSelectionStatus = exports.ApiTextType = exports.ApiBlockType = exports.ApiRelationshipType = void 0;
var base_1 = require("./base");
Object.defineProperty(exports, "ApiRelationshipType", { enumerable: true, get: function () { return base_1.ApiRelationshipType; } });
Object.defineProperty(exports, "ApiBlockType", { enumerable: true, get: function () { return base_1.ApiBlockType; } });
var content_1 = require("./content");
Object.defineProperty(exports, "ApiTextType", { enumerable: true, get: function () { return content_1.ApiTextType; } });
Object.defineProperty(exports, "ApiSelectionStatus", { enumerable: true, get: function () { return content_1.ApiSelectionStatus; } });
var form_1 = require("./form");
Object.defineProperty(exports, "ApiKeyValueEntityType", { enumerable: true, get: function () { return form_1.ApiKeyValueEntityType; } });
var table_1 = require("./table");
Object.defineProperty(exports, "ApiTableEntityType", { enumerable: true, get: function () { return table_1.ApiTableEntityType; } });
Object.defineProperty(exports, "ApiTableCellEntityType", { enumerable: true, get: function () { return table_1.ApiTableCellEntityType; } });
